core/src/main/scala/kafka/server/KafkaRequestHandler.scala

```diff
@@ -53,25 +53,27 @@ object KafkaRequestHandler {
    * Wrap callback to schedule it on a request thread.
    * NOTE: this function must be called on a request thread.
    * @param fun Callback function to execute
+   * @param requestLocal The RequestLocal for the current request handler thread in case we need to call
+   *                     the callback function without queueing the callback request
    * @return Wrapped callback that would execute `fun` on a request thread
    */
-  def wrap[T](fun: T => Unit): T => Unit = {
+  def wrap[T](fun: (RequestLocal, T) => Unit, requestLocal: RequestLocal): T => Unit = {
```

Comment:
Since this is a generic util, it would be useful for future users of this method to understand what kind of `fun` could be passed in. I was thinking that perhaps we could make `fun` of a new interface (sth like `ThreadSafeCallback`). We can document the expectation of `ThreadSafeCallback` (e.g., only executed once, but could be in an arbitrary request handler thread; if an implementation needs to access a share data structure, the access needs to be thread safe, etc). Then, anyone who uses this method in the future needs to be aware of this interface. This may address a bit of Divij's concern. Is that worth doing?

Reply:
That seems reasonable to me.

Reply:
I think ideally in a future change we'd get rid of passing RequestLocal as an argument and maybe make it a static thread local that could be accessed from the point of use rather than being passed through the whole stack.

There are couple problems that contributed to this issue:
- the functions here are already written for asynchronous completion (because we wait for replication) and in such cases generally the convention is that the arguments of a function are not bound to the executing thread (i.e. rooted on the call stack or globally)
- the point of use was outside of the core change so folks didn't look into the specifics of the RequestLocal semantics (i.e. even if appendEntries was an explicit function as it now is, I'm not sure if the problem had been noticed).